snapcraft/internal/lxd.py

```diff
@@ -40,30 +72,104 @@ def __init__(self, snap_output, tar_filename, project_options):
+            data = self._container.files.get(src)
+            destination_file.write(data)
 
     def _container_run(self, cmd):
-        check_call(['lxc', 'exec', self._container_name, '--'] + cmd)
+        if not self._container:
+            raise CalledProcessError(-1, cmd)
+        print('Executing {}'.format(cmd))
+        # API call because self._container.execute doesn't expose IO streams
+        response = self._container.api.exec.post(json={
+            'command': cmd,
+            # Make debconf stop asking for user input
+            'environment': {
+                'DEBIAN_FRONTEND': 'noninteractive',
+            },
+            'wait-for-websocket': True,
+            'interactive': False,
+        })
+        self._fds = response.json()['metadata']['metadata']['fds']
+        operation_id = response.json()['operation'].split('/')[-1]
+        self._parsed = parse.urlparse(
+            self._client.api.operations[operation_id].websocket._api_endpoint)
+        self._manager = WebSocketManager()
+        self._pipe(sys.stdin), self._pipe(sys.stdout), self._pipe(sys.stderr)
+        self._manager.start()
+        while len(self._manager) > 0:
+            sleep(.1)
+        response = self._client.api.operations[operation_id].get()
+        exit_status = response.json()['metadata']['metadata']['return']
+        if exit_status is not 0:
+            raise CalledProcessError(exit_status, cmd)
+
+    def _pipe(self, io):
+        pipe = _CommandWebsocketclient(
+            self._manager, io, self._client.websocket_url)
+        pipe.resource = '{}?secret={}'.format(
+            self._parsed.path, self._fds[str(io.fileno())])
+        pipe.connect()
+        return pipe
+
+    def _get_fingerprint_by_name(self, name):
+        for url in self._client.api.images.get().json()['metadata']:
+            fingerprint = url.split('/')[-1]
+            response = self._client.api.images[fingerprint].get()
+            props = response.json()['metadata']['properties']
+            image_name = '{}:{}/{}'.format(
+                props['os'], props['release'], props['architecture'])
+            if name == image_name:
+                return fingerprint
 
     @contextmanager
     def _create_container(self):
         try:
-            check_call([
-                'lxc', 'launch', '-e',
-                'ubuntu:xenial/{}'.format(self._project_options.deb_arch),
-                self._container_name])
+            fingerprint = self._get_fingerprint_by_name(
+                'ubuntu:xenial/{}'.format(self._project_options.deb_arch))
+            logger.info('Creating container {} with fingerprint {}'.format(
+                self._container_name, fingerprint[:12]))
+            self._container = self._client.containers.create({
+                'name': self._container_name,
+                'ephemeral': True,
+                'source': {
+                    'type': 'image',
+                    'fingerprint': fingerprint,
```

Comment:
Those docs must be outdated. If you try that you'll get "Must specify one of alias, fingerprint or properties for init from image" (https://github.com/lxc/lxd/blob/master/lxd/containers_post.go#L122).